src/cargo/sources/registry/mod.rs

```diff
@@ -812,6 +814,18 @@ impl<'cfg> Source for RegistrySource<'cfg> {
     }
 
     fn block_until_ready(&mut self) -> CargoResult<()> {
+        // Before starting to work on the registry, make sure that
+        // `<cargo_home>/registry` is marked as excluded from indexing and
+        // backups. Older versions of Cargo didn't do this, so we do it here
+        // regardless of whether `<cargo_home>` exists.
+        //
+        // This does not use `create_dir_all_excluded_from_backups_atomic` for
+        // the same reason: we want to exclude it even if the directory already
+        // exists.
+        let registry_base = self.config.registry_base_path();
+        registry_base.create_dir()?;
```

Comment:
Oops. I feel like this kind of `create_dir` operation could make the situation of running cargo on a read-only filesystem worse (see #10096). Maybe we should ignore the error but I am not sure.

Reply:
Oh OK, interesting. I'll look at how that's handled/ignored/tested elsewhere.

Reply:
Actually it looks like this code is already tested in a readonly directory, by `readonly_registry_still_works`.

https://github.com/rust-lang/cargo/blob/f3a444853839a581e56eb70f1e873d39d425724b/tests/testsuite/registry.rs#L2463-L2507

I added an assertion that the path is writable, and it fails in this test.

I think it was OK before because
1. `create_dir` does nothing if the dir exists.
2. `exclude_from_backups_and_indexing` ignores errors.

However I changed it to explicitly drop the result, which seems clearer.

Reply:
Thanks for the analysis!

We should drop `?` here. And perhaps `git_fs.create_dir()` in `GitSource::block_until_ready` also needs this trick.

```patch
- let _ = registry_base.create_dir()?;
+ let _ = registry_base.create_dir();
```

Reply:
Oh of course, sorry.

Reply:
done